### PR TITLE
feat: add file output for TypeScript generator

### LINKED ISSUE #399: Generating TypeScript models directly to files
#### Reason/Context
As a precursor to #172 we need to give the user the option to generate models directly to files.

This is related to #288 as the full output is needed to write it to a file.

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR adds file output for the TypeScript generator.

**Related issue(s)**
Fixes #399

### CODE REVIEW

File: src/generators/typescript/TypeScriptFileGenerator.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+export class TypeScriptFileGenerator extends TypeScriptGenerator implements AbstractFileGenerator<TypeScriptRenderCompleteModelOptions> {
+  /**
+   * Generates all the models to an output directory each model with their own separate files. 
+   * 
+   * @param input
+   * @param outputDirectory where you want the models generated to
+   * @param options
+   */
+  public async generateToFiles(input: Record<string, unknown> | CommonInputModel, outputDirectory: string, options?: TypeScriptRenderCompleteModelOptions): Promise<OutputModel[]> {
```
Comment (magicmatatjahu) [inline src/generators/typescript/TypeScriptFileGenerator.ts:15]: I wrote about that in PR related to the JS File Renderer, but I'll write about that also here.

Have in mind that someone will want to render models with ESM not CJS module system, so you `TypeScriptRenderCompleteModelOptions` should have option to choose it - by default we can use ESM.

**Review by magicmatatjahu — CHANGES_REQUESTED** [review]

File: src/generators/typescript/TypeScriptFileGenerator.ts (hunk 1 above)
Comment (jonaslagoni) [reply src/generators/typescript/TypeScriptFileGenerator.ts:15]: Yep, changing 👍

**Comment by jonaslagoni** [comment]
Blocked by #557

**Comment by jonaslagoni** [comment]
I really hate windows and Linux differences.......... 🙄 

`tsc --downlevelIteration -t es5 ${path.resolve(renderOutputPath, './*.ts')}` works fine in linux... However, windows only works with `tsc --downlevelIteration -t es5 --baseUrl ${renderOutputPath}`....

File: src/generators/typescript/TypeScriptGenerator.ts [hunk 2]
```diff
@@ -48,16 +47,37 @@ export class TypeScriptGenerator extends AbstractGenerator<TypeScriptOptions,Typ
...
     const outputModel = await this.render(model, inputModel);
-    const modelDependencies = model.getNearestDependencies().map((dependencyModelName) => {
-      const formattedDependencyModelName = this.options.namingConvention?.type ? this.options.namingConvention.type(dependencyModelName, { inputModel, model: inputModel.models[String(dependencyModelName)] }) : dependencyModelName;
-      return `import ./${formattedDependencyModelName};`;
+    let modelDependencies = model.getNearestDependencies();
+    //Ensure model dependencies have their rendered name
+    modelDependencies = modelDependencies.map((dependencyModelName) => {
+      return this.options.namingConvention?.type ? this.options.namingConvention.type(dependencyModelName, { inputModel, model: inputModel.models[String(dependencyModelName)] }) : dependencyModelName;
+    });
+    //Filter out any dependencies that is recursive to it'self
```
Comment (magicmatatjahu) [inline src/generators/typescript/TypeScriptGenerator.ts:57]:
```suggestion
    //Filter out any dependencies that is recursive to itself
```

**Review by magicmatatjahu — APPROVED** [review]

**Review by magicmatatjahu — APPROVED** [review]

**Comment by jonaslagoni** [comment]
Thanks @magicmatatjahu !

**Pull request merged**